cardano-cli/src/Cardano/CLI/Run/Debug/CheckNodeConfiguration.hs

```diff
@@ -0,0 +1,188 @@
+import qualified Data.Aeson.KeyMap as Aeson
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as LBS
+import           Data.List (isSuffixOf)
+import           Data.Maybe (catMaybes)
+import qualified Data.Text as Text
+import qualified Data.Text.IO as Text
+import qualified Data.Yaml as Yaml
+import           System.Directory (removeFile)
+import           System.FilePath (takeDirectory, (</>))
+
+runCheckNodeConfig :: CheckNodeConfigCmdArgs -> ExceptT DebugCmdError IO ()
+runCheckNodeConfig (CheckNodeConfigCmdArgs configFile fixTheFile) = do
+  nodeBS <- liftIO $ BS.readFile configFilePath
+  nodeConfig :: NodeConfig <- Yaml.decodeThrow nodeBS
+  -- We need both forms, because NodeConfig doesn't cover all fields of the node configuration file.
+  -- If we were to serialize it back, we would loose data.
+  nodeConfigValue :: Aeson.Value <- Yaml.decodeThrow nodeBS
+  nodeConfigObj <-
+    case nodeConfigValue of
+      Aeson.Object obj -> pure obj
+      _ -> throwE $ DebugNodeConfigReadCmdError configFilePath "Expected an Object, but got something else"
+  analyses <- catMaybes <$> mapM (\era -> mkAnalysis era configFile nodeConfig) eras
+  nodeConfigObj' <- foldM (runAnalysis configFile fixTheFile) nodeConfigObj analyses
+  when (nodeConfigObj' /= nodeConfigObj) $ do
+    -- Write the fixed node configuration file
+    -- Better to remove the file first, to avoid a file busy exception when writing
+    liftIO $ removeFile configFilePath
+    -- Cheap trick to try to preserve the input file's format
+    let looksLikeJson = ".json" `isSuffixOf` configFilePath
+    liftIO $
+      if looksLikeJson
+        then LBS.writeFile configFilePath $ Aeson.encodePretty nodeConfigObj'
+        else Yaml.encodeFile configFilePath nodeConfigObj'
+  liftIO $ putStrLn $ "Successfully checked node configuration file: " <> configFilePath
+ where
+  configFilePath = unFile configFile
+  eras :: [AnyCardanoEra] = [minBound .. maxBound]
+
+data Analysis = Analysis
+  { concernedEra :: AnyCardanoEra
+  , genesisFilepath :: FilePath
+  -- ^ The path to the genesis file, as found in the node configuration file, if any
+  , actualHash :: Text.Text
+  -- ^ The actual hash of the genesis file, computed from the file's content
+  , expectedHash :: Maybe Text.Text
+  -- ^ The hash of the genesis file, as found in the node configuration file, if any.
+  }
+
+runAnalysis
+  :: NodeConfigFile 'In
+  -- ^ The node configuration file. It's not read by this function, but used for producing error messages.
+  -> FixNodeConfigFile
+  -> Aeson.Object
+  -- ^ The node configuration file, as a JSON object. It will get written back to disk.
+  -> Analysis
+  -> ExceptT DebugCmdError IO Aeson.Object
+  -- ^ The fixed node configuration file
+runAnalysis configFile fixMode nodeConfigObj (Analysis{concernedEra = era@(AnyCardanoEra cEra), genesisFilepath = _, actualHash, expectedHash}) = do
+  case (expectedHash, fixMode) of
```

Comment:
I think the fix mode is not worth it. We have all this additional clunky logic to insert into maps etc. Reporting the errors is enough and if the hash does not turn out to be the expected hash the user is probably going to take the time to investigate why vs just fixing it automatically.

Reply:
Removed the `fix` mode :+1: